### ISSUE TITLE
Bug: [explicit-function-return-type] false positive on defaulting parameters

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#8950]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.4.3&fileType=.ts&code=C4TwDgpgBAwghgGwQITgYwNZQLxQBQCUOAfFAG4D2AlgCYDcAUAwGYCuAdmsFRe1M3gDmFYGgAWcAFyxEKdFlyESUAN5QAvgWmVaqpuqA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6RAD2MbMroDNYTMvkoB7JrWiJ8saJKKl0URNGhjokcGAC%2BIHUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```TypeScript
type CallBack = () => void;

function f(gotcha: CallBack = () => { }): void {

}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/explicit-function-return-type": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

I expect no error since CallBack makes it a typed function expression

### Actual Result

"Missing return type on function":
```ts
function f(gotcha: CallBack = () => { }): void {
                                 ^^
```
